java-squid/src/main/java/org/sonar/java/ast/parser/JavaParser.java

```diff
@@ -39,4 +45,26 @@ private JavaParser() {
       JavaLexer.COMPILATION_UNIT);
   }
 
+  @Override
+  public Tree parse(File file) {
+    return createParentLink((JavaTree) super.parse(file));
+  }
+
+  @Override
+  public Tree parse(String source) {
+    return createParentLink((JavaTree) super.parse(source));
+  }
+
+  private static Tree createParentLink(JavaTree parent) {
+    if (!parent.isLeaf()) {
+      for (Iterator<Tree> iter = parent.childrenIterator(); iter.hasNext();) {
+        JavaTree next = (JavaTree) iter.next();
+        if (next != null) {
+          next.setParent(parent);
```

Comment:
So, for trees "sharing" children (which is already known to be problematic on our side), we will end up with a single parent all the time, which is not a real representation of the hierarchy...

With the following code ```final int a = 0, b = 0;```, I would have expected the `ModifierKeywordTree` corresponding to `final`, or the `TypeTree` corresponding to `int`, to have two parents.

Consequently, if you try something like in the following code snippet, you will face strange behavior. Given the `VariableTree` b.

```
VariableTree a = b.modifiers().parent(); // when you would expect to get back to b
```